src/peerbook/libp2p_peer.erl

```diff
@@ -186,14 +314,35 @@ decode_list(Bin) ->
 -spec decode(binary()) -> peer().
 decode(Bin) ->
     Msg = libp2p_peer_pb:decode_msg(Bin, libp2p_signed_peer_pb),
-    verify(Msg).
+    true = verify(Msg),
+    Msg.
+
 
-%% @doc Cryptographically verifies a given peer.
--spec verify(peer()) -> peer().
-verify(Msg=#libp2p_signed_peer_pb{peer=Peer=#libp2p_peer_pb{}, signature=Signature}) ->
+%% @doc Cryptographically verifies a given peer and it's
+%% associations. Returns trus if the given peer can be verified or
+%% throws an error if the peer or one of it's associations can't be
+%% verified
+-spec verify(peer()) -> true.
+verify(Msg=#libp2p_signed_peer_pb{peer=Peer=#libp2p_peer_pb{associations=Assocs}, signature=Signature}) ->
     EncodedPeer = libp2p_peer_pb:encode_msg(Peer),
-    PubKey = libp2p_crypto:address_to_pubkey(Peer#libp2p_peer_pb.address),
+    PubKey = libp2p_crypto:address_to_pubkey(address(Msg)),
     case public_key:verify(EncodedPeer, sha256, Signature, PubKey) of
-        true -> Msg;
+        true ->
+            lists:all(fun({_AssocType, #libp2p_association_list_pb{associations=AssocEntries}}) ->
+                              MsgAddress = address(Msg),
+                              lists:all(fun(Assoc) ->
+                                                association_verify(Assoc, MsgAddress)
+                                        end, AssocEntries)
+                      end, Assocs);
         false -> error(invalid_signature)
     end.
+
+%%
+%% Internal
+%%
+
+-spec sign_peer(#libp2p_peer_pb{}, libp2p_crypto:sig_fun()) -> peer().
+sign_peer(Peer, SigFun) ->
+    EncodedPeer = libp2p_peer_pb:encode_msg(Peer),
```

Comment:
Is this the signature over the whole peer record? Won't this invalidate the signature if some aspect of the peer changes that is not really germane to whether we control this key? Why not just sign the p2p address?

Reply:
hmm.. we sign the whole peer record.. we always did.. since an association adds to the overall peer record we need to sign it again? Were we signing peerbook entries wrong the whole time?

Reply:
Sorry, yes, I misread this change as where we were signing using the associated key, disregard.